### PR TITLE
description: Explain questions

### PR DESCRIPTION
**Pull request by teodutu**

### CODE REVIEW

**Comment by RazvanN7** [comment]
@teodutu is this supposed to be the entire section on quizzes or is it just a preview?

As a preview, it's ok, but as an entire section it needs more meat.

**Comment by teodutu** [comment]
It's a preview that I plan to turn into the complete thing before it's merged.

File: description.md [hunk 1]
```diff
@@ -14,13 +13,48 @@ Each topic in this file contains a small description and links to the files cove
...
+
+Is this a sample question?
+
+## Question Answers
+
++ Yes
+
+- No
+
+- Maybe
```
Comment (Alex-deVis) [inline description.md:47]: Don't forget about the feedback.

**Comment by teodutu** [comment]
@RazvanN7 I've added the "meat". Give this another look when you can.

**Review by Alex-deVis — APPROVED** [review]
LGTM

File: description.md [hunk 2]
```diff
@@ -17,6 +17,86 @@ Each section details how a specific type of content is to be developed, stored a
...
+Quizzes may be part of practical sessions, lectures or standalone as forms of evaluation.
+
+#### Purpose
+
+Quizzes provide a means to continuously evaluate students.
+This creates a positive feedback loop between trainers and trainees whereby both can track the students' progress in real time.
+
+Furthermore, quizzes serve as practice items for students to solve at their own pace.
+They may also be used for evaluation items that contribute to the students' overall grade.
+
```
Comment (RazvanN7) [inline description.md:33]: For completeness, I think it is worth mentioning the downsides of quizzes:

- not suitable for some fields of study
- binary feedback loop: correct/false
- maybe others?

File: description.md [hunk 3]
```diff
@@ -17,6 +17,86 @@ Each section details how a specific type of content is to be developed, stored a
...
+It supports graded assignments, quizzes, storing teaching materials and many other educational activities.
+
+- [SmileyTutor (formerly called TutorWeb)]((https://tutor-web.net/)) is another free learning platform whose highlight is that it was built around the [SmyleyCoin cryptocurrency](#cryptocurrency-rewards) as a reward for completing drills and quizzes.
+
+- [tech.io](https://tech.io/) is another quiz platform.
+More simplistic than Moodle or SmileyTutor, it allows teachers to easily create _playgrounds_, which are learning environments designed around specific subjects, made up of reading material, followed by practical exercises and quizzes.
+
+Each of these platforms supports a different syntax for uploading questions.
+To make our quizzes comply with educational platforms easily, we need to store them in a generic format that is easily representable and automatically convertible to other required formats.
+For this reason, we are storing questions in Markdown format.
```
Comment (RazvanN7) [inline description.md:52]: For this reason and for the sake of consistency (we use Markdown EVERYWHERE).

File: description.md [hunk 4]
```diff
@@ -17,6 +17,86 @@ Each section details how a specific type of content is to be developed, stored a
...
+We are developing a modular and easily extendible converter that ports quizzes from Markdown to other formats.
```
Comment (RazvanN7) [inline description.md:53]: I would rephrase this to: "The format that we use is easily convertible to other formats. We do not implement converters for all the existing formats, however, we do provide a program that is easily extendible to support any kind of format."

File: description.md [hunk 5]
```diff
@@ -17,6 +17,86 @@ Each section details how a specific type of content is to be developed, stored a
...
+- No
+
+- Maybe
+
+## Feedback
+
+The question is self-reflecting and its title is "Sample Question Name", so it is a sample question.
+```
+
+Note that correct answers are marked as list entries starting with `+`, while incorrect ones are marked with `-`.
```
Comment (RazvanN7) [inline description.md:98]: Maybe you should add a few sentences about the fact that the `#*` sections are actually keys and are not predefined, therefore you can add any kind of entry (that is why it is extensible). Also, it would be nice to talk about the the program that creates the dictionary which then can be used to output the desired format.

**Review by RazvanN7 — COMMENTED** [review]
Nice job! Mind the comments.

File: description.md [hunk 6]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+Quizzes are questions that verify specific information and must be verifiable automatically.
+Therefore, multiple-choice questions are regarded as quizzes, while open-ended ones are not.
+Quizzes may be part of practical sessions, lectures or standalone as forms of evaluation.
+
+#### Purpose
+
+Quizzes provide a means to continuously evaluate students.
+This creates a positive feedback loop between trainers and trainees whereby both can track the students' progress in real time.
+
+Furthermore, quizzes serve as practice items for students to solve at their own pace.
```
Comment (razvand) [inline description.md:31]: Reference drills again here, saying something like "together with drills".

File: description.md [hunk 7]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+In addition, non-technical fields of study may find little benefit in quizzes as they rely more on arguments essays rather than concrete, technical information.
+
+#### Deployment
+
+Each trainer uses a specific teaching rendering platform.
+Most platforms support quizzes in various formats, such as whether they are graded or ungraded, whether the answers are multiple choice or text boxes, etc.
+Out of the existing teaching platforms, a few highlights are:
+
+- [Moodle](https://moodle.org/) is a free open-source learning platform dedicated to course management.
+It supports graded assignments, quizzes, storing teaching materials and many other educational activities.
```
Comment (razvand) [inline description.md:49]:
```suggestion
  It supports graded assignments, quizzes, storing teaching materials and many other educational activities.
```

File: description.md [hunk 8]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+
+- [SmileyTutor (formerly called TutorWeb)]((https://tutor-web.net/)) is another free learning platform whose highlight is that it was built around the [SmyleyCoin cryptocurrency](#cryptocurrency-rewards) as a reward for completing drills and quizzes.
+
+- [tech.io](https://tech.io/) is another quiz platform.
+More simplistic than Moodle or SmileyTutor, it allows teachers to easily create _playgrounds_, which are learning environments designed around specific subjects, made up of reading material, followed by practical exercises and quizzes.
+
```
Comment (razvand) [inline description.md:55]: I would add [Kahoot](https://kahoot.it/).

File: description.md [hunk 9]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+Each of these platforms supports a different syntax for uploading questions.
+To make our quizzes comply with educational platforms easily, we need to store them in a generic format that is easily representable and automatically convertible to other required formats.
+For this reason and for the sake of consistency, we are storing questions in the same Markdown format as all of our written material.
+
+We are developing a modular and easily extendible converter that ports quizzes from Markdown to other formats.
+The format that we use to store quizzes is easily convertible to other formats.
+Therefore, we do not implement converters for all the existing formats.
+Instead, we do provide a program that is easily extendible to support any kind of format.
```
Comment (razvand) [inline description.md:63]: This is not clear. What do we provide and what we don't provide?

File: description.md [hunk 10]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+
+Another desired aspect is that quizzes be easily extractable and convertible between formats, as described [below](#deployment), and repurposed by other trainers.
+
+Despite all of their strengths, quizzes also come with a few shortcomings.
+One such downside is the binary nature of the feedback loop.
+Because multiple-choice answers may be either correct or wrong, there is little room for nuance regarding the feedback that students receive upon answering incorrectly.
+Similarly, quizzes cannot be easily tailored and personalized for the individual needs of each student.
+In addition, non-technical fields of study may find little benefit in quizzes as they rely more on arguments essays rather than concrete, technical information.
+
+#### Deployment
```
Comment (razvand) [inline description.md:42]: Before this, add a summary of steps related to quizzes: brainstorm the question, write it down using the OEH format, store in the proper location in the repository, convert it to the required publisher format, publish / deploy it.

I would then detail each step in a subsection.

File: description.md [hunk 11]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+
+## Feedback
+
+The question is self-reflecting and its title is "Sample Question Name", so it is a sample question.
+```
+
+Note that correct answers are marked as list entries starting with `+`, while incorrect ones are marked with `-`.
+In addition, the headings starting with `#` are just keys for different sections of the quiz.
+Therefore, none of them are mandatory given our format and the developer is free to add or remove any of them.
+
```
Comment (razvand) [inline description.md:111]: Detail the structure of the `quizzes/` directory as part of each topic.

**Review by razvand — CHANGES_REQUESTED** [review]
I did a first round of review. I generally like how it's structured. It needs a bit more concreteness on certain areas. I will do a 2nd round after you apply the review.

File: description.md [hunk 12]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+Unlike more complex formats like LaTeX, Markdown offers a much simpler syntax which makes quizzes easier to contribute to by educators and students alike.
+
+The format we propose relies on headers to represent categories such as difficulty, topics, question text, feedback.
+Each question has a title written on the first line as a a level 1 header (`#`).
+Each category is specified as a level 2 header (`##`).
+The contents of the category consist of the text beneath the header.
+For multiple-choice questions, correct answers are marked as list entries starting with `+`, while incorrect ones are marked with `-`.
+
+See the example below:
+
```
Comment (StefanJum) [inline description.md:77]: Will every question be in it's own file?
If not, we should add that questions in the same file will be separated by 2 empty lines. If we will have one question per file we should also add that.

File: description.md [hunk 13]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+- No
+
+- Maybe
+
+## Feedback
+
+The question is self-reflecting and its title is "Sample Question Name", so it is a sample question.
+```
+
+Note that correct answers are marked as list entries starting with `+`, while incorrect ones are marked with `-`.
```
Comment (StefanJum) [inline description.md:108]: Add that you can have more then one correct answer.

File: description.md [hunk 14]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+In addition, the headings starting with `#` are just keys for different sections of the quiz.
+Therefore, none of them are mandatory given our format and the developer is free to add or remove any of them.
+
+#### Converter
+
+Given the format above, we are providing a converter that can reformat quizzes from one format to another.
+It is bidirectional, so it can translate from Markdown to Moodle XML, for example, and vice-versa.
```
Comment (StefanJum) [inline description.md:115]: Add a link to the [Moodle XML format](https://docs.moodle.org/402/en/Moodle_XML_format)

**Review by StefanJum — CHANGES_REQUESTED** [review]

File: description.md [hunk 15]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
 
+#### Definition
+
+Quizzes are questions that verify specific information and must be verifiable automatically.
+Therefore, multiple-choice questions are regarded as quizzes, while open-ended ones are not.
+Quizzes may be part of practical sessions, lectures or standalone as forms of evaluation.
+
+#### Purpose
+
+Quizzes provide a means to continuously evaluate students.
```
Comment (RazvanN7) [inline description.md:28]:
```suggestion
Quizzes provide a means to evaluate student understanding.
```

File: description.md [hunk 16]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+They may also be used for evaluation items that contribute to the students' overall grade.
+
+Another desired aspect is that quizzes be easily extractable and convertible between formats, as described [below](#deployment), and repurposed by other trainers.
+
+Despite all of their strengths, quizzes also come with a few shortcomings.
+One such downside is the binary nature of the feedback loop.
+Because multiple-choice answers may be either correct or wrong, there is little room for nuance regarding the feedback that students receive upon answering incorrectly.
+Similarly, quizzes cannot be easily tailored and personalized for the individual needs of each student.
+In addition, non-technical fields of study may find little benefit in quizzes as they rely more on arguments essays rather than concrete, technical information.
+
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply description.md:41]: Yes, I would also link it to the last sentence. Something along the lines of: "However, even for these fields, quizzes may be used to evaluate concrete knowledge of specific information".

File: description.md (hunk 9 above)
Comment (RazvanN7) [reply description.md:63]: I think that an image may help here.

File: description.md (hunk 12 above)
Comment (RazvanN7) [reply description.md:77]: Each quiz in its own file. Makes it  simpler for future use.

File: description.md [hunk 17]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+- Maybe
+
+## Feedback
+
+The question is self-reflecting and its title is "Sample Question Name", so it is a sample question.
+```
+
+Note that correct answers are marked as list entries starting with `+`, while incorrect ones are marked with `-`.
+In addition, the headings starting with `#` are just keys for different sections of the quiz.
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply description.md:109]: It's mandatory, but it's still a key in a hashtable. We probably need to discuss the exact format of the hashtable.

File: description.md [hunk 18]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+Therefore, none of them are mandatory given our format and the developer is free to add or remove any of them.
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply description.md:110]: I don't know if we should do that. I mean, having any kind of string as a key offers flexibility. Maybe someone will want to use the headings "Question, Answer, Correct". The converter shouldn't enforce any of these. It's up to the linter associated to the repo to enforce such rules.

File: description.md [hunk 19]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+#### Converter
+
+Given the format above, we are providing a converter that can reformat quizzes from one format to another.
+It is bidirectional, so it can translate from Markdown to Moodle XML, for example, and vice-versa.
+In addition, it is also capable of translating between non-Markdown formats, such as Moodle XML and SmileyTutor.
+If a new quiz format is desired, the converter may be easily extendable to accommodate it and convert between it and any other format.
+
+The converter achieves this level of flexibility by internally representing each quiz as a dictionary, from which any other format can be derived.
+The keys of this dictionary are the `##` headings and the value of each key is the text beneath the heading.
+This also allows the converter to support a JSON format for free, as it is directly mapped to the converter's internal representation.
```
Comment (RazvanN7) [inline description.md:121]: Maybe link to what a json is? Theoretically, non-technical people might read this too.

File: description.md (hunk 7 above)
Comment (teodutu) [reply description.md:49]: What's the effect of this? I see no difference in the rendered output. Is it just to align this sentence with the one above?

File: description.md [hunk 20]
```diff
@@ -17,6 +17,121 @@ Each section details how a specific type of content is to be developed, stored a
...
+For example, the sample question above will be represented as a JSON as follows:
+
+```json
+"Sample Question Name": {
+    "Difficulty": "easy",
+    "Tags": ["topic-1", "topic-2"],
+    "Question Text": "Is this a sample question?",
+    "Question Answers": ["Yes", "No", "Maybe"],
+    "Correct Answer": "Yes",
```
[the remark this answers is not in the document]
Comment (teodutu) [reply description.md:130]: We need to think about the grade part. How do we add this to the Markdown file? All this information kind of leads to having a separate subsection for each answer. @razvand @RazvanN7 ?

[the remark this answers is not in the document]
Comment (StefanJum) [reply description.md:130]: You can have the grade of a correct answer just be `1 / nr of correct answers`. About the wrong answers, yes, we need to either add it in the md file or have it as an option to the questions parser.

File: description.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (gstefans) [reply description.md:41]: I'm not sure I understand this discussion:  in SmileyTutor (former tutor-web) we allocate different drill items (quiz questions) depending on student ability, so allocation depends on where they are in the learning process. This qualifies as personalised learning, yet using multiple-choice drill items. Note that these are not just ordinary quizzes as in an exam, but individual multiple choice questions (=items in item theory).

So I don't see how this is correct: "Similarly, quizzes cannot be easily tailored and personalized for the individual needs of each student."

Next, this "beause multiple-choice answers may be either correct or wrong, there is little room for nuance regarding the feedback that students receive upon answering incorrectly." --- in SmileyTutor we normally have a detailed explanation which the student sees once they have submitted the answer... In principle one could have different explanations depending on which incorrect answer is submitted. So how is this a problem with multiple-choice??

[the remark this answers is not in the document]
Comment (RazvanN7) [reply description.md:41]: @gstefans The discussion is more about non-technical fields such as philosophy, psychology etc. where having quizzes is not ideal. But this could apply to mathematics also: you have much more insights on the understanding of a student if you evaluate how the student got to the result than if you just evaluate the answer.

Maybe this is not well phrased, but we want to highlight that multiple choice questions may not be ideal in specific scenarios.

[the remark this answers is not in the document]
Comment (gstefans) [reply description.md:41]: Right. I still think some care in wording is needed:

Questions with open-ended answer options certainly have many benefits over items with multiple-choice answers. But it is not factually correct that there is no flexibility in providing personalised learning with multiple-choice items. 

It is also a common claim that multiple-choice items only provide for rote learning, but this is also not factually correct - we have papers demonstrating that appropriate use of such items in fact provides transferable knowledge...

[the remark this answers is not in the document]
Comment (RazvanN7) [reply description.md:41]: @teodutu please rephrase this so that we do not have any factually incorrect statements.

File: description.md (hunk 10 above)
Comment (teodutu) [reply description.md:42]: Isn't this too much detail? The steps themselves are self-explanatory. For example, the whole section on Quizzes explains how to store them in the OEH format, so I don't think we should go into that here.

File: description.md (hunk 11 above)
Comment (teodutu) [reply description.md:111]: What do you mean by "each topic"?

File: description.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (teodutu) [reply description.md:109]: No, @StefanJum is right. It was a typo on my side. The `#` heading is just the title. I have no idea how to handle it in the converter.

File: description.md (hunk 7 above)
Comment (Alex-deVis) [reply description.md:65]: There is no difference, it produces nesting if you also add an empty line before.

File: description.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (teodutu) [reply description.md:38]: > there is little room for nuance regarding the feedback that students receive upon answering incorrectly

I meant to say this is a con of automatic grading. It's difficult to find the exact step that the student got wrong and insist on that.

File: description.md [hunk 21]
```diff
@@ -17,6 +17,157 @@ Each section details how a specific type of content is to be developed, stored a
...
+There may be multiple correct answers.
+In addition, the headings starting with `##` are just keys for different sections of the quiz.
+Therefore, none of them are mandatory given our format and the developer is free to add or remove any of them.
+
+#### Converter
+
+Given the format above, we are providing a converter that can reformat quizzes from one format to another.
+It is bidirectional, so it can translate from Markdown to [Moodle XML](https://docs.moodle.org/402/en/Moodle_XML_format), for example, and vice-versa.
+In addition, it is also capable of translating between non-Markdown formats, such as [Moodle XML](https://docs.moodle.org/402/en/Moodle_XML_format) and SmileyTutor.
+If a new quiz format is desired, the converter may be easily extendable to accommodate it and convert between it and any other format.
```
Comment (Alex-deVis) [inline description.md:144]: This sounds repetitive to me.

File: description.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (Alex-deVis) [reply description.md:157]: > We need to think about the grade part. How do we add this to the Markdown file?

You do not care about the grading part in the context of a single quiz. It has its correct answer(s) and that is it. Grading comes into place when you have a succession of quizzes.

File: description.md [hunk 22]
```diff
@@ -17,6 +17,157 @@ Each section details how a specific type of content is to be developed, stored a
...
+```
+
+To use this converter, a person is required to run the command below.
+The script will infer the fact that the input format is Markdown and the output is Moodle XML from the extensions of the input files (`.md` and `.mxml`).
+
+```console
+python3 quiz_converter.py convert -i input_file.md -o output_file.mxml
+```
+
+The converter may also be run by the builder if the teacher desires to integrate quizzes within their published course and not use a third party hosting website, such as those outlined [above](#deployment).
```
Comment (Alex-deVis) [inline description.md:169]: This is vague. Is it run automatically or it is up to the teacher to configure it in the builder?

File: description.md [hunk 23]
```diff
@@ -17,6 +17,157 @@ Each section details how a specific type of content is to be developed, stored a
...
+    "Difficulty": "easy",
+    "Tags": ["topic-1", "topic-2"],
+    "Question Text": "Is this a sample question?",
+    "Question Answers": ["Yes", "No", "Maybe"],
+    "Correct Answer": "Yes",
+    "Feedback": "The question is self-reflecting and its title is \"Sample Question Name\", so it is a sample question."
+}
+```
+
+To use this converter, a person is required to run the command below.
```
Comment (Alex-deVis) [inline description.md:162]: Consider highlighting this with a section e.g. `#### Usage`. Even if it is a short one, is something that people might search for.

File: description.md (hunk 22 above)
Comment (teodutu) [reply description.md:169]: Not sure whether the builder should provide this functionality off-the-shelf or be left up to the teacher.

File: description.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (razvand) [reply description.md:271]: What @Alex-deVis said. Its part of the implementation. Not the storing of content.
